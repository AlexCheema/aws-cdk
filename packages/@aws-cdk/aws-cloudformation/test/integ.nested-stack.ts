import sns = require('@aws-cdk/aws-sns');
import { App, CfnParameter, Construct, Stack } from '@aws-cdk/core';
import cfn = require('../lib');

interface MyNestedStackProps {
  readonly topicCount: number;
  readonly topicNamePrefix: string;
}

class MyNestedStack extends cfn.NestedStack {
  constructor(scope: Construct, id: string, props: MyNestedStackProps) {
    const topicNamePrefixLogicalId = 'TopicNamePrefix';

    super(scope, id, {
      parameters: {
        [topicNamePrefixLogicalId]: props.topicNamePrefix
      }
    });

    const topicNamePrefixParameter = new CfnParameter(this, 'TopicNamePrefix', { type: 'String' });

    for (let i = 0; i < props.topicCount; ++i) {
      new sns.Topic(this, `topic-${i}`, { displayName: `${topicNamePrefixParameter.valueAsString}-${i}`});
    }
  }
}

class MyTestStack extends Stack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new MyNestedStack(this, 'NestedStack1', { topicCount: 3, topicNamePrefix: 'Prefix1'});
    new MyNestedStack(this, 'NestedStack2', { topicCount: 2, topicNamePrefix: 'Prefix2' });
  }
}

const app = new App();
new MyTestStack(app, 'nested-stacks-test');
app.synth();