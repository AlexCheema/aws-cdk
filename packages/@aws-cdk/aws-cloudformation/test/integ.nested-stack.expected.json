{
  "Parameters": {
    "NestedStack1NestedStackAssetS3Bucket5FED0F84": {
      "Type": "String",
      "Description": "S3 bucket for asset \"nested-stacks-test/NestedStack1.NestedStack/Asset\""
    },
    "NestedStack1NestedStackAssetS3VersionKeyD7FDFFF0": {
      "Type": "String",
      "Description": "S3 key for asset version \"nested-stacks-test/NestedStack1.NestedStack/Asset\""
    },
    "NestedStack1NestedStackAssetArtifactHash1F5040A3": {
      "Type": "String",
      "Description": "Artifact hash for asset \"nested-stacks-test/NestedStack1.NestedStack/Asset\""
    },
    "NestedStack2NestedStackAssetS3BucketF8C8921A": {
      "Type": "String",
      "Description": "S3 bucket for asset \"nested-stacks-test/NestedStack2.NestedStack/Asset\""
    },
    "NestedStack2NestedStackAssetS3VersionKey0C402EEC": {
      "Type": "String",
      "Description": "S3 key for asset version \"nested-stacks-test/NestedStack2.NestedStack/Asset\""
    },
    "NestedStack2NestedStackAssetArtifactHash4D4982F5": {
      "Type": "String",
      "Description": "Artifact hash for asset \"nested-stacks-test/NestedStack2.NestedStack/Asset\""
    }
  },
  "Resources": {
    "NestedStack1NestedStackNestedStack1NestedStackResource7E64AD52": {
      "Type": "AWS::CloudFormation::Stack",
      "Properties": {
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region"
              },
              ".",
              {
                "Ref": "AWS::URLSuffix"
              },
              "/",
              {
                "Ref": "NestedStack1NestedStackAssetS3Bucket5FED0F84"
              },
              "/",
              {
                "Fn::Select": [
                  0,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "NestedStack1NestedStackAssetS3VersionKeyD7FDFFF0"
                      }
                    ]
                  }
                ]
              },
              {
                "Fn::Select": [
                  1,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "NestedStack1NestedStackAssetS3VersionKeyD7FDFFF0"
                      }
                    ]
                  }
                ]
              }
            ]
          ]
        },
        "Parameters": {
          "TopicNamePrefix": "Prefix1"
        }
      }
    },
    "NestedStack2NestedStackNestedStack2NestedStackResourceFDF82E43": {
      "Type": "AWS::CloudFormation::Stack",
      "Properties": {
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region"
              },
              ".",
              {
                "Ref": "AWS::URLSuffix"
              },
              "/",
              {
                "Ref": "NestedStack2NestedStackAssetS3BucketF8C8921A"
              },
              "/",
              {
                "Fn::Select": [
                  0,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "NestedStack2NestedStackAssetS3VersionKey0C402EEC"
                      }
                    ]
                  }
                ]
              },
              {
                "Fn::Select": [
                  1,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "NestedStack2NestedStackAssetS3VersionKey0C402EEC"
                      }
                    ]
                  }
                ]
              }
            ]
          ]
        },
        "Parameters": {
          "TopicNamePrefix": "Prefix2"
        }
      }
    }
  }
}